src/test/java/com/bobocode/blyznytsia/bibernate/persister/EntityPersisterImplTest.java

```diff
@@ -0,0 +1,192 @@
+      assertThrows(PersistenceException.class, () -> entityPersister.findOneBy(SampleEntity.class, keyField, "val1"));
+    }
+  }
+
+  @Nested
+  class FindById {
+    @Test
+    void returnsEntityById() {
+      var expectedEntity = new SampleEntity(1L, "val1");
+      var actualEntityOptional = entityPersister.findById(SampleEntity.class, 1L);
+      assertTrue(actualEntityOptional.isPresent());
+      assertEquals(expectedEntity, actualEntityOptional.get());
+    }
+
+    @Test
+    void returnsEmptyOptionalWhenNoValueFound() {
+      var nonExistingEntityOptional = entityPersister.findById(SampleEntity.class, -1L);
+      assertTrue(nonExistingEntityOptional.isEmpty());
+    }
+
+    @Test
+    void throwsIllegalArgumentExceptionWhenIdIsNull() {
+      assertThrows(IllegalArgumentException.class, () -> entityPersister.findById(SampleEntity.class, null));
+    }
+
+  }
+
+  @Nested
+  class Insert {
+    @Test
+    void persistsDataInTheDatabase() {
+      SampleEntity sampleEntity = new SampleEntity(null, "val 4");
+      entityPersister.insert(sampleEntity);
+      var entityFromDatabaseOptional = getEntityFromDB(sampleEntity.getId());
+      assertTrue(entityFromDatabaseOptional.isPresent());
+      assertEquals(sampleEntity, entityFromDatabaseOptional.get());
+    }
+
+    @Test
+    void setsEntityGeneratedId() {
+      SampleEntity sampleEntity = new SampleEntity(null, "val 4");
+      entityPersister.insert(sampleEntity);
+      assertNotNull(sampleEntity.getId());
+    }
+
+    @Test
+    void wrapsSQLExceptions() throws NoSuchFieldException {
+      var nonExistingEntity = new NonExistingEntity();
+      var exception = assertThrows(PersistenceException.class, () -> entityPersister.insert(nonExistingEntity));
+      assertTrue(exception.getCause() instanceof SQLException);
+    }
+  }
+
+  @Nested
+  class Update {
+    @Test
+    void changesAnEntityInTheDatabase() {
+      var entityToUpdate = new SampleEntity(1001L, "valueAfterUpdate");
+      entityPersister.update(entityToUpdate);
+      var entityFromDBOptional = getEntityFromDB(1001L);
+      assertTrue(entityFromDBOptional.isPresent());
+      assertEquals(entityToUpdate, entityFromDBOptional.get());
+    }
+
+    @Test
+    void throwsPersistenceExceptionWhenEntityIsTransient() {
+      var entityToUpdate = new SampleEntity(null, null);
+      assertThrows(PersistenceException.class, () -> entityPersister.update(entityToUpdate));
+    }
+  }
+
+  @Nested
+  class Delete {
+    @Test
+    void deletesEntity() {
+      var entityToDelete = new SampleEntity(1002L, null);
+      entityPersister.delete(entityToDelete);
+      assertTrue(getEntityFromDB(1002L).isEmpty());
+    }
+
+    @Test
+    void throwsPersistenceExceptionWhenEntityIsTransient() {
+      var entityToDelete = new SampleEntity(null, null);
+      assertThrows(PersistenceException.class, () -> entityPersister.delete(entityToDelete));
+    }
+  }
+
+  private Optional<SampleEntity> getEntityFromDB(Long id) {
```

Comment:
We can use `entityPersister.findOneBy` but it will negatively impact test isolation: in such case, if `findOneBy` fails, it will trigger cascade failures of the dependent test.

Reply:
okay, I am fine with it